### DISCUSSION TITLE
Test app: Markdown Combiner

### DISCUSSION THREAD

**Discussion by LeonMatthesKDAB** [KDAB/cxx-qt#371] [Show and tell]
Markdown combiner using CXX-Qt.

I'm currently developing a simple Markdown combiner app (target: Linux + Android) using CXX-Qt.

The goal is to create an app that can work cross-platform to combine multiple markdown checklists into one.
In this discussion I'll list issues I've encountered and the reasoning to why I need them.

This can hopefully guide our future issue prioritization.

**Comment by LeonMatthesKDAB** [comment]
#55 

I need to convert a QUrl, which is returned from a QtQuick FolderDialog to a local file name, which is certainly possible, though quite cumbersome to do in Rust atm.

**Reply by LeonMatthesKDAB** [reply]
Imagined API:
``` rust
impl QUrl {
    pub fn to_local_file(&self) -> Option<QString>;
}
```

**Reply by ahayzen-kdab** [reply]
Right QUrl is missing loads of the methods that would be useful. This would just be bound to https://doc.qt.io/qt-6/qurl.html#toLocalFile `QString QUrl::toLocalFile() const`. If we don't do the Option and follow the Qt API  it would be a straight CXX bridge definition with no  wrapper methods.

**Comment by LeonMatthesKDAB** [comment]
#372

Caused a binding loop in my QML (See example in issue).

Current workaround:
Expose a `#[qinvokable]` update_value() that does the equality check (requires conversion to Rust atm.) and then calls the setter.

**Comment by LeonMatthesKDAB** [comment]
#373
Minor nuisance, can do `use X;` from within invokable to get around this.

**Comment by LeonMatthesKDAB** [comment]
#374
Not critical atm, but might come in handy in future.

**Comment by LeonMatthesKDAB** [comment]
#375
Not really pressing, just something I noticed.

**Comment by LeonMatthesKDAB** [comment]
#53 
Especially the Debug impl would have been really nice!
Then just using `#[derive(Debug)]` would have been possible :partying_face:

**Comment by LeonMatthesKDAB** [comment]
#42 
Especially the ability to define my own setters would have been very helpful.
Right now, I can't just assign the property from QML, as that will break things.
My current workaround is similar to https://github.com/KDAB/cxx-qt/discussions/371#discussioncomment-4332429 .
Define an invokable that does the actual value setting and updates all other related state.

**Reply by ahayzen-kdab** [reply]
Also for the workaround assume you have realised you could derive from the Rust class in C++ to add the property?

Eg have a read only field private to Rust, then an invokable to access it (could even have a signal to notify change)

```rust
#[cxx_qt::qobject]
pub struct MyObject {
  read_only_field: i32
}

impl qobject::MyObject {
   #[qinvokable] 
   pub fn read_only_field(&self) -> i32 { self.read_only_field() }
}
```

Then in C++
```cpp
class MyObjectWithProp : public MyObject {
  Q_OBJECT
  Q_PROPERTY(::std::int32_t readOnlyField READ readOnlyField CONSTANT)
...
};
```

Then register and use MyObjectWithProp in QML with the `readOnlyField` :-)

The same process can be done to setup your own setters / signals etc.

**Reply by LeonMatthesKDAB** [reply]
Good point, I was trying to see what CXX-Qt can do without working around its restrictions too much...

**Comment by LeonMatthesKDAB** [comment]
#476

Associated constants are not allowed.
